### ISSUE TITLE
Large file transfer hangs

### ISSUE THREAD

**Issue by deckar01** [jedisct1/piknik#18]
**Server:** Xubuntu 17.10 (i7-860 16GB)
**Client:** MacOS 10.12.6 (i7-4750 8GB)
**Piknik:** v0.9.1 (protocol version: 5)

Copying a 5.3GB file from the client spawns a piknik process on the client that consumes about 10GB and hangs. The server memory usage seems to only grow the expected ~5.3GB. Copying on the server produces the content that was previously on the clipboard (not the 5.3GB file).

**Comment by jedisct1** [comment]
Hi!

Piknik was not really designed to transfer large files.

It wouldn't take much to extend the protocol to transfer large blobs in smaller chunks; it's just that it's not a use case I had in mind when writing it and you're the first person to ask about that kind of usage. But if more people are interested in using Piknik to transfer large files, this can be added.

Still, as long as you have enough memory on the server and on the client, you should be able to transfer files that fit. But you may have to tweak the `maxLenMb` (maximum length in megabytes) and `dataTimeout (maximum time to complete a transfer)` settings, using the relevant command line switches.

**Comment by deckar01** [comment]
I though I had enough free memory on the client and the server for the file (6GB & 10GB). There was no messaging to indicate that anything was happening. I assumed the memory constraint would be on the server, but the client seems to need free memory for double the file size. That constraint seems like it might be a bug. Verbose logging or a progress bar might help.

**Comment by jedisct1** [comment]
Maybe it was just stuck transferring the data ?

**Comment by jedisct1** [comment]
Yes. That can be worked around, though. We can definitely encrypt in-place.

**Comment by deckar01** [comment]
I added logging and let the process run to completion. Reading the 5.3GB file resulted in 13.45GB of memory consumption. This is expected behavior from the go GC according to https://github.com/golang/go/issues/16269.
 Encrypting the file increased this to 18.50GB. Signing has no real impact on memory usage.

The program spends a couple seconds trying to send the data, then the `writer.Flush()` call errors with the message `<nil>`. The error handling is actually broken on that line. It is using the `err` variable from a previous assignment instead of the flush call. Fixing that shows `write tcp (client_ip):61409->(server_ip):8075: write: broken pipe`.

Running a sanity check on a small test message shows that piknik is configured correctly.

The whole process takes about 5 minutes, which means the read, encryption, and signing operations are averaging about 54MB/s. I suspect a significant amount of this time is spent paging virtual memory on and off the client SSD, because the process only spent about 2 minutes in user space.

It looks like the server might be closing the socket before the client is done preparing the data. As far as I can tell, the server seems to be using the data timeout (which defaults to 1 hour). I suspect there is some built in default timeout that is closing the socket due to a lack of communication for a prolonged period of time. I will see if I can add some logging to the server and find out why it is giving up so soon.

**Comment by jedisct1** [comment]
That's still a lot of memory :/

```go
    ciphertext := ciphertextWithNonce[24:]
    cipher.XORKeyStream(ciphertext, content)
```

We should really just use a single buffer here. Enlarge `content`, shift the data 24 bytes to the right, and encrypt in-place. That would drastically reduce memory usage.

**Comment by deckar01** [comment]
I was able to shave the run time down to 2m50s by stating the file size, allocating the nonce + cipher text buffer once, reading the file into the buffer, and encrypting in place. The memory usage is still reporting about 15GB of usage, but considering the significant speedup, it seems to just be spaced reserved by Go that is never used. Here is a break down of where the time is going now:

- Read: 20s
- Encryption: 24s
- Signature: 126s

That signature seems really slow. Ed25519 is supposed to have really fast signing. I am tempted to try the libsodium-go binding and see how much faster that is.

The pipe is still broken when I try to write. I tried different file sizes and noticed that the connectiong is closing after 10 seconds of data processing. That is supposed to be piknik's connection timeout not it's data timeout. I suspect the data timeout needs to be explicitly set after the initial connection is made to avoid a dropped connection while the client is busy processing the data.
